src/main/java/com/project/bookapp/service/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,113 @@
+import com.project.bookapp.exceptions.EntityNotFoundException;
+import com.project.bookapp.mapper.CartItemMapper;
+import com.project.bookapp.mapper.ShoppingCartMapper;
+import com.project.bookapp.model.Book;
+import com.project.bookapp.model.CartItem;
+import com.project.bookapp.model.ShoppingCart;
+import com.project.bookapp.model.User;
+import com.project.bookapp.repository.book.BookRepository;
+import com.project.bookapp.repository.cart.ShoppingCartRepository;
+import com.project.bookapp.repository.item.CartItemRepository;
+import com.project.bookapp.repository.user.UserRepository;
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+@Transactional
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final UserService userService;
+    private final BookRepository bookRepository;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final UserRepository userRepository;
+
+    @Override
+    public ShoppingCartResponseDto save(Long id, CartItemRequestDto cartItemRequestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(id)
+                .orElseGet(() -> createNewShoppingCart(id));
+
+        Book book = bookRepository.findById(cartItemRequestDto.getBookId())
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Book not found with ID: "
+                                + cartItemRequestDto.getBookId()));
+
+        CartItem cartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> item.getBook().getId().equals(cartItemRequestDto.getBookId()))
+                .findFirst()
+                .orElseGet(() -> {
+                    CartItem newItem = cartItemMapper.toEntity(cartItemRequestDto);
+                    newItem.setShoppingCart(shoppingCart);
+                    newItem.setBook(book);
+                    return newItem;
+                });
+
+        cartItem.setQuantity(cartItem.getQuantity() + cartItemRequestDto.getQuantity());
+
+        CartItem savedCartItem = cartItemRepository.save(cartItem);
+        shoppingCart.addCartItem(savedCartItem);
+
+        shoppingCartRepository.save(shoppingCart);
+
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto findById(Long id) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(id)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Shopping cart with id " + id + " not found"));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    public ShoppingCart createNewShoppingCart(Long userId) {
```

Comment:
add method to interface ShoppingCartService
```suggestion
    @Override
    public ShoppingCart createNewShoppingCart(User user) {
```